File: problems/583_delete_operation_for_two_strings/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    int minDistance(string word1, string word2) {
        int m = word1.size(), n = word2.size();
        vector<int> dp(n + 1, 0);
        for (int i = 1; i <= m; i++) {
            vector<int> newDp(n + 1, 0);
            for (int j = 1; j <= n; j++) {
                if (word1[i-1] == word2[j-1]) newDp[j] = dp[j-1] + 1;
                else newDp[j] = max(dp[j], newDp[j-1]);
            }
            dp = newDp;
        }
        return m + n - 2 * dp[n];
    }
};

File: problems/583_delete_operation_for_two_strings/solution.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        m, n = len(word1), len(word2)
        dp = [0] * (n + 1)
        for i in range(1, m + 1):
            new_dp = [0] * (n + 1)
            for j in range(1, n + 1):
                if word1[i - 1] == word2[j - 1]:
                    new_dp[j] = dp[j - 1] + 1
                else:
                    new_dp[j] = max(dp[j], new_dp[j - 1])
            dp = new_dp
        lcs = dp[n]
        return m + n - 2 * lcs
